language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java

```diff
@@ -151,14 +169,27 @@ public SonarLintLanguageServer(InputStream inputStream, OutputStream outputStrea
     this.logOutput = new LanguageClientLogOutput(client);
 
     backgroundProcess = launcher.startListening();
+
+    this.logger = loggerFactory.apply(this.client);
+    this.engineCache = engineCacheFactory.apply(logOutput, logger);
+    this.serverInfoCache = new ServerInfoCache(logger);
   }
 
-  public static SonarLintLanguageServer bySocket(int port, Collection<URL> analyzers) throws IOException {
+  static SonarLintLanguageServer bySocket(int port, Collection<URL> analyzers) throws IOException {
     Socket socket = new Socket("localhost", port);
-    return new SonarLintLanguageServer(socket.getInputStream(), socket.getOutputStream(), analyzers);
+
+    BiFunction<LanguageClientLogOutput, ClientLogger, EngineCache> engineCacheFactory = (logOutput, logger) -> {
+      StandaloneEngineFactory standaloneEngineFactory = new StandaloneEngineFactory(analyzers, logOutput, logger);
+      ConnectedEngineFactory connectedEngineFactory = new ConnectedEngineFactory(logOutput, logger);
+      return new DefaultEngineCache(standaloneEngineFactory, connectedEngineFactory);
+    };
+
+    Function<SonarLintLanguageClient, ClientLogger> loggerFactory = client -> new DefaultClientLogger(client);
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Replace this lambda with a method reference. ([squid:S1612](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1612))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWNEPxp56NykrOWnW1Uv&open=AWNEPxp56NykrOWnW1Uv)

[ISSUE_KEY]: AWNEPxp56NykrOWnW1Uv